### PR TITLE
Add JAX coupling

### PR DESCRIPTION
**Pull request by nbouziani**
This PR adds support to couple JAX and Firedrake in a differentiable fashion. It extends the previous work on coupling Firedrake and PyTorch (cf. #3394 and #2804) to support the JAX framework as well as downstream libraries. It allows to embed JAX model into Firedrake, thanks to the `MLOperator` external operator class, and to embed Firedrake system into JAX. These new capabilities can be obtained by importing from the `firedrake.ml.jax` module. It also adds options to install JAX on the fly when installing Firedrake. Main additions:

- Add a JAX custom operator (`FiredrakeJaxOperator`) to represent Firedrake operators expressed as a ReducedFunctional, wherein forward and backward computations are delegated to the reduced functional.
- Add a new ml operator (`JaxOperator`) to represent JAX operations embedded in Firedrake
- Add support for installing JAX
- Add mapping operator from JAX to Firedrake, and vice-versa
- Some tests

### CODE REVIEW

File: firedrake/ml/jax/fem_operator.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    F : pyadjoint.ReducedFunctional
+        The reduced functional to wrap.
+
+    Returns
+    -------
+    firedrake.ml.jax.fem_operator.FiredrakeJaxOperator
+        A JAX custom operator that wraps the reduced functional `F`.
+    """
+    # TODO: Add citation
+    # Citations().register(TODO)
```
Comment (nbouziani) [inline firedrake/ml/jax/fem_operator.py:115]: This will come shortly.

File: firedrake/ml/jax/fem_operator.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+    if not isinstance(F, ReducedFunctional):
+        raise ValueError("F must be a ReducedFunctional")
+
+    jax_op = FiredrakeJaxOperator(F)
+    return partial(FiredrakeJaxOperator.forward, jax_op)
```
Comment (nbouziani) [inline firedrake/ml/jax/fem_operator.py:121]: Add a comment to explain why this workaround exists

**Review by JDBetteridge — CHANGES_REQUESTED** [review]
Please use type hinting as per the "new" policy:
https://github.com/firedrakeproject/firedrake/wiki/Docstrings

**Review by JDBetteridge — APPROVED** [review]
Docstrings and function signatures have been updated to use type hinting.

**Pull request merged**